compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmPropertiesLowering.kt

```diff
@@ -67,6 +67,18 @@ class JvmPropertiesLowering(private val backendContext: JvmBackendContext) : IrE
         return expression
     }
 
+    private fun isDefaultAccessorForCompanionPropertyBackingFieldOnCurrentClass(
+        property: IrProperty,
+        function: IrSimpleFunction
+    ): Boolean {
+        if (function.origin != IrDeclarationOrigin.DEFAULT_PROPERTY_ACCESSOR) return false
+        if (property.isLateinit) return false
+        if ((currentFunction?.irElement as? IrFunction)?.isInline == true) return false
```

Comment:
I'm not sure that lambda expression is already extracted to separate class before this lowering (or it could be changed in future), it's also good to have next cases in tests:
```
class A {
   ...
    inline fun g(crossinline s: () -> Unit) {
        {
            s()
            v = "OK"            
        } ()
    }

    inline fun g2(crossinline s: () -> Unit) {
        object {
            fun run() {
                s()
                v = "OK"
            }
        }.run ()
    }
}```

Reply:
This did indeed expose another case that is not working. The code snippet above is currently not problematic. Both the object and the lambda are represented as classes and therefore the `currentClass` check will filter them out.

However, accesses from a LOCAL_FUNCTION_FOR_LAMBDA in an inline function are not counted as being in an inline function and therefore end up using the accessibility bridge instead of the getting on the Companion object. That is not good. I added a test and changed the code to filter out everything that is nested inside of an inline function and not just the functions itself. I incorrectly assumed that there would be no local functions inside an inline function (as local functions are rejected by the frontend), however, lambdas at this point are represented as nested functions.

This still leads to differences with the current backend. For the following example:

```
class A {
  companion object {
    val x = 32
  }

  inline fun f(g: () -> Unit) {
    g()
  }

  fun h() {
    f {
      x
    }
  }
}
```

The access to x in h will use direct property access. That happens on the current backend as well.

The difference is that with the current JVM_IR code, the same is true if the lambda is marked cross inline, in that case the accessibility bridge will be used from the $$inlined class. That should be fine for binary compatibility, I think, as this code belongs on this class and therefore if we change the companion object this code will be recompiled as well. But, what do you think @max-kammerer?

Reply:
> The difference is that with the current JVM_IR code, the same is true if the lambda is marked cross inline, in that case the accessibility bridge will be used from h. That should be fine for binary compatibility, I think, as this code belongs on this class and therefore if we change the companion object this code will be recompiled as well. But, what do you think @max-kammerer?

Yes, it's safe.
Thank you also for adding similar cases with `inline h` )